### PR TITLE
Preserve parent OCI config

### LINKED ISSUE #4415: OCI JSON config gets lost
## Version of Singularity:

master branch

### Expected behavior

Parent image envs are preserved.

### Actual behavior

Parent image envs get lost.

### Steps to reproduce behavior

First, pull from docker:
```
$ singularity pull docker://quay.io/wtsicgp/cgppindel:hotfix_v3.2.1
$ singularity sif dump 2 cgppindel_hotfix_v3.2.1.sif
{"User":"ubuntu","Env":["PATH=/opt/wtsi-cgp/bin:/opt/wtsi-cgp/biobambam2/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","OPT=/opt/wtsi-cgp","PERL5LIB=/opt/wtsi-cgp/lib/perl5","LD_LIBRARY_PATH=/opt/wtsi-cgp/lib","LC_ALL=en_US.UTF-8","LANG=en_US.UTF-8"],"Cmd":["/bin/bash"],"WorkingDir":"/home/ubuntu","Labels":{"description":"cgpPindel docker","maintainer":"cgphelp@sanger.ac.uk","uk.ac.sanger.cgp":"Cancer, Ageing and Somatic Mutation, Wellcome Trust Sanger Institute","version":"v3.2.1"}}
```
All ok, envs are there.

Now push/pull SIF:
```
$ singularity push cgppindel_hotfix_v3.2.1.sif library://sashayakovtseva/default/test:envs
$ singularity pull library://sashayakovtseva/default/test:envs
$ singularity sif dump 2 test_envs.sif 
{"User":"ubuntu","Env":["PATH=/opt/wtsi-cgp/bin:/opt/wtsi-cgp/biobambam2/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin","OPT=/opt/wtsi-cgp","PERL5LIB=/opt/wtsi-cgp/lib/perl5","LD_LIBRARY_PATH=/opt/wtsi-cgp/lib","LC_ALL=en_US.UTF-8","LANG=en_US.UTF-8"],"Cmd":["/bin/bash"],"WorkingDir":"/home/ubuntu","Labels":{"description":"cgpPindel docker","maintainer":"cgphelp@sanger.ac.uk","uk.ac.sanger.cgp":"Cancer, Ageing and Somatic Mutation, Wellcome Trust Sanger Institute","version":"v3.2.1"}}
```
Looks good so far, envs still there.

Now build from the image:
```
$ cat > env.def <<EOF
Bootstrap: localimage
From: test_envs.sif

%runscript
    env

%post
    echo "Hello from inside the container"
    echo "Install additional software here"
EOF

$ sudo singularity build env.sif env.def
$ singularity sif list env.sif 
Descriptor list:
ID   |GROUP   |LINK    |SIF POSITION (start-end)  |TYPE
------------------------------------------------------------------------------
1    |1       |NONE    |32768-32927               |Def.FILE
2    |1       |NONE    |36864-140537856           |FS (Squashfs/*System/amd64)
```

The JSON config partition is lost!

### PR DESCRIPTION
**Pull request by sashayakovtseva**
**Description of the Pull Request (PR):**

In order to preserve parent's OCI config SIF packer now stores it (if any) into a bundle's JSON objects map.

Also minor changes:
- add a constant for oci config JSON object 
- change methods to functions where receiver is not used
- use sif magic as defined in sif package
- use nil slice instead of `[]string{}` in squashfs extractor
- remove redundant uint64 conversion
- do not check rootfs descriptor datatype b/c it is already checked in the next called function
- use camelCase/common names where possible
- add periods in godoc comments

**This fixes or addresses the following GitHub issues:**

- Fixes #4415 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/pkg/build/sources/packer_sif_linux.go [hunk 1]
```diff
@@ -76,10 +76,22 @@ func (p *SIFPacker) unpackSIF(b *types.Bundle, srcfile string) (err error) {
 		return fmt.Errorf("unrecognized partition format")
 	}
 
+	ociReader, err := image.NewSectionReader(img, "oci-config.json", -1)
+	if err != nil && err != image.ErrNoSection {
```
Comment (cclerget) [inline internal/pkg/build/sources/packer_sif_linux.go]: Maybe something like :

```
	if err == image.ErrNoSection {
                sylog.Debugf("No oci-config.json section found")
        } else if err != nil {
                return fmt.Errorf("could not read image OCI config: %v", err)
	} else {
                ociConfig, err := ioutil.ReadAll(ociReader)
                ...
        }
```

Other errors returned by `NewSectionReader` are fatal in this case

File: pkg/image/sif.go [hunk 2]
```diff
@@ -89,9 +85,6 @@ func (f *sifFormat) initializer(img *Image, fileinfo os.FileInfo) error {
 		if !desc.Used {
 			continue
 		}
-		if desc.Datatype != sif.DataPartition {
```
Comment (cclerget) [inline pkg/image/sif.go:92]: What the reason behind this removal ?

**Review by cclerget — CHANGES_REQUESTED** [review]

File: pkg/image/sif.go (hunk 2 above)
Comment (sashayakovtseva) [reply pkg/image/sif.go:92]: The next call to `desc.GetPartType()` already does this check:

```

// GetPartType extracts the Parttype field from the Extra field of a Partition Descriptor
func (descr *Descriptor) GetPartType() (Parttype, error) {
	if descr.Datatype != DataPartition {
		return -1, fmt.Errorf("expected DataPartition, got %v", descr.Datatype)
	}
...
```

File: internal/pkg/build/sources/packer_sif_linux.go (hunk 1 above)
Comment (sashayakovtseva) [reply internal/pkg/build/sources/packer_sif_linux.go]: I thought we might want to count those error as non fatal, but now I think that this mean SIF is corrupted. Agree, thanks.

**Review by mem — APPROVED** [review]

**Review by ikaneshiro — APPROVED** [review]
lgtm, thanks

**Review by cclerget — APPROVED** [review]

**Pull request merged**